flocker/node/agents/ebs.py

```diff
@@ -683,6 +701,87 @@ def _is_cluster_volume(cluster_id, ebs_volume):
     return False
 
 
+def _attach_volume_and_wait_for_device(
+    volume, attach_to,
+    attach_volume, detach_volume,
+    device, attach_attempt,
+    blockdevices,
+):
+    """
+    Attempt to attach an EBS volume to an EC2 instance and wait for the
+    corresponding OS device to become available.
+
+    :param BlockDeviceVolume volume: The Flocker representation of the volume
+        to attach.
+    :param unicode attach_to: The EC2 instance id to which to attach it.
+    :param attach_volume: A function like ``EC2Connection.attach_volume``.
+    :param attach_volume: A function like ``EC2Connection.detach_volume``.
+    :param unicode device: The OS device path to which to attach the device.
+    :param int attach_attempt: The number of this attempt to attach the volume
+        (counting from 1).  This is used to decide whether it's worth
+        continuing to retry or not.
+    :param list blockdevices: The OS device paths which are already present on
+        the system before this operation is attempted (primarily useful to make
+        testing easier).
+
+    :raise: Anything ``attach_volume`` can raise.  Or
+        ``AttachedUnexpectedDevice`` if the volume appears to become attached
+        to the wrong OS device file.
+
+    :return: ``True`` if the volume is attached and accessible via the expected
+        OS device file.  ``False`` if the attempt times out without succeeding.
```

Comment:
Hmm.  Mixed feelings about this.  On one hand, yes this code is gross and mixes some layers.  I did intentionally keep the exception handling down here because I don't want to introduce new exception handling around the `detach_volume` call which is what happens if I do the obvious refactoring to move this up.

While thinking about this I did implement something that introduces a couple new exception types for the control flow necessary to put the retry logic up a layer but make sure it still only applies to attach failures.  At the end of that exercise I realized the reason I did the refactoring at all was to avoid changing logic in a function I couldn't unit test - which is exactly what the next layer up is.

So I'm inclined to add tests for the retry logic where it is now.  The downside is that the layering remains kind of gross.  The upside is I can have unit tests for it.